src/zonetree/in_memory/builder.rs

```diff
@@ -0,0 +1,172 @@
+///
+/// `ZoneBuilder` is used to specify the content of a single zone one,
+/// resource record or RRset at a time, and to then turn that specification
+/// into a populated in-memory [`Zone`].
+///
+/// <div class="warning">
+///
+/// Already have a zonefile in [presentation format]?
+///
+/// Check out the example [module docs] which shows how to use
+/// [`inplace::Zonefile`], [`parsed::Zonefile`] and `ZoneBuilder` together
+/// without having to manually insert each resource record into the
+/// `ZoneBuilder` yourself.
+///
+/// </div>
+///
+/// Each `ZoneBuilder` builds a single zone with a named apex and a single
+/// [`Class`]. All resource records within the zone are considered to have the
+/// specified class.
+///
+/// `ZoneBuilder` has dedicated functions for inserting certain kinds of
+/// resource record properly into the zone in order to cater to RR types that
+/// require or benefit from special handling when [`ReadableZone::query()`] is
+/// invoked for the zone.
+///
+/// # Usage
+///
+/// To use `ZoneBuilder`:
+/// - Call [`ZoneBuilder::new()`] to create a new builder.
+/// - Call the various `insert_()` functions to add as many resource records
+/// as needed.
+/// - Call [`ZoneBuilder::build()`] to exchange the builder for a populated
+///   [`Zone`].
+///
+/// [module docs]: crate::zonetree
+/// [`inplace::Zonefile`]: crate::zonefile::inplace::Zonefile
+/// [`parsed::Zonefile`]: crate::zonefile::parsed::Zonefile
+/// [presentation format]:
+///     https://datatracker.ietf.org/doc/html/rfc9499#section-2-1.16.1.6.1.3
+/// [`ReadableZone::query()`]: crate::zonetree::ReadableZone::query()
+pub struct ZoneBuilder {
+    apex: ZoneApex,
+}
+
+impl ZoneBuilder {
+    /// Creates a new builder for the specified apex name and class.
+    ///
+    /// All resource records in the zone will be considered to have the
+    /// specified [`Class`].
+    #[must_use]
+    pub fn new(apex_name: StoredDname, class: Class) -> Self {
+        ZoneBuilder {
+            apex: ZoneApex::new(apex_name, class),
+        }
+    }
+
+    /// Builds an in-memory [`Zone`] from this builder.
+    ///
+    /// Calling this function consumes the [`ZoneBuilder`]. The returned
+    /// `Zone` will be populated with the resource records that were inserted
+    /// into the builder.
+    #[must_use]
+    pub fn build(self) -> Zone {
+        Zone::new(self.apex)
+    }
+
+    /// Inserts a related set of resource records.
+    ///
+    /// Inserts a [`SharedRrset`] for the given owner name.
+    pub fn insert_rrset(
+        &mut self,
+        name: &impl ToDname,
+        rrset: SharedRrset,
+    ) -> Result<(), OutOfZone> {
+        match self.get_node(self.apex.prepare_name(name)?) {
+            Ok(node) => node.rrsets().update(rrset, Version::default()),
+            Err(apex) => apex.rrsets().update(rrset, Version::default()),
+        }
+        Ok(())
+    }
+
+    /// Insert one or more resource records that represent a zone cut.
```

Comment:
What this actually does is convert a node (or owner name?) into a zone cut – it doesn’t actually insert RRsets into the RRsets of the node.

Maybe the name of the method is poorly chosen.

(This also applies to `insert_cname` below.)

Reply:
Does it even convert? It actually sort of does "insert" because if the "special" part of the node is empty it will push a value into it.

I don't think the caller really needs to know what it does internally, from the caller perspective the operations on a zone builder insert high level zone file entities into the zone. Perhaps one could argue that it doesn't insert "resource records" into its own structure, but from a caller perspective it takes resource records (in the form of `Rrset` and `SharedRrset` and `StoredRecord`) and "inserts" them into the zone builder, i.e. does something with them/processes them/whatever.

I don't have a better name or description at the moment and as you say if we change this maybe we should change the other fns too. It's a vague term but we could change them to `process_xxx` instead of `insert_xxx`?

Reply:
Added to the todo list.